test/compiler/irpasses.jl

```diff
@@ -1881,3 +1881,33 @@ end
 @test !fully_eliminated() do
     all(iszero, Iterators.repeated(0))
 end
+
+## Test that cfg_simplify respects implicit `unreachable` terminators
+let code = Any[
+        # block 1
+        GotoIfNot(Core.Argument(2), 4),
+        # block 2
+        Expr(:call, Base.throw, "error"), # an implicit `unreachable` terminator
+        # block 3
+        Expr(:call, :opaque),
+        # block 4
+        ReturnNode(nothing),
+    ]
+    ir = make_ircode(code; ssavaluetypes=Any[Union{}, Union{}, Any, Union{}])
+
+    # Unfortunately `compute_basic_blocks` does not notice the `throw()` so it gives us
+    # the wrong CFG. Instead manually construct the intended CFG:
```

Comment:
I wouldn't say the CFG is wrong. It's valid to have extra unreachable bits at the end of the basic block.

Reply:
Right, it's not an invalid CFG - It's just not the (more precise) one we need for this test.

I'll clarify in the comment

Reply:
```suggestion
    # a slightly imprecise CFG. Instead manually construct the CFG we need for this test:
```

Reply:
Thinking on this again, unfortunately I think this CFG _is_ wrong in general, for the same reasons we encountered in https://github.com/JuliaLang/julia/pull/53512. Since Enter/Exit nodes add structural conditions to the CFG, spurious fallthroughs are invalid in general.

I think we might be safe from hitting this in practice because we end up not having these implicit must-throw terminators in fresh code going into inference, but nonetheless we probably need to fix that...